README.md
# radarchart

[![Build Status](https://travis-ci.org/MangoTheCat/radarchart.svg?branch=master)](https://travis-ci.org/MangoTheCat/radarchart) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/MangoTheCat/radarchart?branch=master&svg=true)](https://ci.appveyor.com/project/MangoTheCat/radarchart) [![CRAN Status](http://www.r-pkg.org/badges/version/radarchart)](http://www.r-pkg.org/pkg/radarchart)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/radarchart)](http://www.r-pkg.org/pkg/radarchart)
[![Total CRAN downloads](http://cranlogs.r-pkg.org/badges/grand-total/radarchart)](http://www.r-pkg.org/pkg/radarchart)



An R implementation of the radar chart from the [chart.js](http://www.chartjs.org/) javascript library.

## Installation

You can install from CRAN.
```r
install.packages("radarchart")
```

To install the latest version direct from GitHub you'll need devtools installed. Assuming you have this run:
```r
devtools::install_github("MangoTheCat/radarchart")
```

Or if you want to hack about with the code then clone the repository, change directory into it and run
```r
devtools::install()
```

Note: `htmlwidgets` packages don't work well with `devtools::load_all()`.

## Usage

Once installed you can make a radar chart

```r
library(radarchart)

labs <- c("Communicator", "Data Wangler", "Programmer",
          "Technologist",  "Modeller", "Visualizer")

scores <- list(
  "Rich" = c(9, 7, 4, 5, 3, 7),
  "Andy" = c(7, 6, 6, 2, 6, 9),
  "Aimee" = c(6, 5, 8, 4, 7, 6)
)

chartJSRadar(scores = scores, labs = labs, maxScale = 10)
```
<img src="inst/figs/readmePlot01.jpg" alt="Static image of example output"  />
<p class="caption">Static version. Real plots are interactive</p>

Alternatively, you may supply a data frame and `chartJSRadar` will pickup the labels from the first column of the data. Also we're showing an option to display data set labels in the mouse over.

```r
scores <- data.frame("Label"=c("Communicator", "Data Wangler", "Programmer",
                     "Technologist",  "Modeller", "Visualizer"),
                     "Rich" = c(9, 7, 4, 5, 3, 7),
                     "Andy" = c(7, 6, 6, 2, 6, 9),
                     "Aimee" = c(6, 5, 8, 4, 7, 6))

chartJSRadar(scores, maxScale = 10, showToolTipLabel=TRUE)
```
<img src="inst/figs/readmePlot02.jpg" alt="Static image of example output"  />
<p class="caption">Static version. Real plots are interactive</p>

You can now also add a title

```r
chartJSRadar(skills, main = "Data Science Radar")
```
<img src="inst/figs/readmePlot03.jpg" alt="Static image of example output"  />
<p class="caption">Static version. Real plots are interactive</p>

## Shiny

As it's based on htmlwidgets it's easy to use `radarchart` with Shiny. Just use the 

```r
chartJSRadarOutput("ID", width = "450", height = "300")
```
function in your `ui.R` and call `chartJSRadar` as normal in your `server.R`. A minimal example can be found in [`inst/shiny-examples/basic`](https://github.com/MangoTheCat/radarchart/tree/master/inst/shiny-examples/basic). You can run this with the `runExampleApp` function

```r
runExampleApp("basic")
```

An `"options"` app is available to help construct more customised radar charts.

## License

MIT © Mango Solutions, Nick Downie


NEWS.md
### 0.3.0

- Upgrade to chart.js v2.4.0
- Allows titles with `main` option
- Allows legends

### 0.2.0

- Plots with no data will render (with no data) [issue 5](https://github.com/MangoTheCat/radarchart/issues/5)
- Change Shiny defaults so the canvas isn't huge [issue 3](https://github.com/MangoTheCat/radarchart/issues/3)
- Added basic shiny example
- Added options shiny example
- runExample function to easily access the shiny apps

### 0.1.2

Fix test error with new htmlwidgets version

## 0.1.1

First CRAN release

### 0.0.2

- Data frame input accepted.
- colMatrix now aligned to `col2rgb` output (transpose of previous version).

### 0.0.1

First GitHub release.


R/chartJSRadar.R
#' Make a ChartJS Radar Plot
#'
#' R bindings to the radar plot in the chartJS library
#'
#' @param scores Data frame or named list of scores for each axis. 
#' If \code{labs} is not specified then labels are taken from the first column (or element).
#' @param labs Labels for each axis. If left unspecified labels are taken from 
#' the scores data set. If set to NA then labels are left blank.
#' @param width Width of output plot
#' @param height Height of output plot
#' @param main Character: Title to be displayed
#' @param maxScale Max value on each axis
#' @param scaleStepWidth Spacing between rings on radar
#' @param scaleStartValue Value at the centre of the radar
#' @param responsive Logical. whether or not the chart should be responsive and resize when the browser does
#' @param labelSize Numeric. Point label font size in pixels
#' @param showLegend Logical whether to show the legend
#' @param addDots Logical. Whether to show a dot for each point
#' @param colMatrix Numeric matrix of rgb colour values. If \code{NULL} defaults are used
#' @param polyAlpha Alpha value for the fill of polygons
#' @param lineAlpha Alpha value for the outlines
#' @param showToolTipLabel Logical. If \code{TRUE} then data set labels are shown in the tooltip hover over
#' @param ... Extra options passed straight to chart.js. Names must match existing options
#' \url{http://www.chartjs.org/docs/#getting-started-global-chart-configuration}
#'
#' @import htmlwidgets
#' @import htmltools
#'
#' @export
#' @examples 
#' # Using the data frame interface
#' chartJSRadar(scores=skills)
#' 
#' # Or using a list interface
#' labs <- c("Communicator", "Data Wangler", "Programmer", "Technologist",  "Modeller", "Visualizer")
#'
#' scores <- list("Rich" = c(9, 7, 4, 5, 3, 7),
#'  "Andy" = c(7, 6, 6, 2, 6, 9),
#'  "Aimee" = c(6, 5, 8, 4, 7, 6))
#'
#' # Default settings
#' chartJSRadar(scores=scores, labs=labs)
#' 
#' # Fix the max score
#' chartJSRadar(scores=scores, labs=labs, maxScale=10)
#' 
#' # Fix max and spacing
#' chartJSRadar(scores=scores, labs=labs, maxScale=12, scaleStepWidth = 2)
#' 
#' # Change title and remove legend
#' chartJSRadar(scores=scores, labs=labs, main = "Data Science Radar", showLegend = FALSE)
#' 
#' # Add pass through settings for extra options
#' chartJSRadar(scores=scores, labs=labs, maxScale =10, scaleLineWidth=5)
#' 
chartJSRadar <- function(scores, labs, width = NULL, height = NULL, main = NULL,
                         maxScale = NULL, scaleStepWidth = NULL,
                         scaleStartValue = 0, responsive = TRUE, labelSize = 18,
                         showLegend = TRUE,
                         addDots = TRUE, colMatrix = NULL, polyAlpha = .2,
                         lineAlpha = .8, showToolTipLabel = TRUE, ...) {
  
  # Should we keep variable names consistent from chart.js to R?
  # Then we can just pass through anything that doesn't need preprocessing
  
  # If it comes through as a single vector wrap it back up
  if(!is.list(scores)) scores <- list(scores)

  # This block trys to handle different ways the data might
  # not look how we expect. Missings etc.
  if(missing(labs)) {
    if(is.character(scores[[1]]) | is.factor(scores[[1]])) {
      labs <- scores[[1]] # Copy the first column of scores into the labels
      if (length(scores) > 1) { # This catches empty data
        scores[[1]] <- NULL # Drop the labels column
      } else {
        # Add a dummy column with no data
        scores <- data.frame("null"=rep(NA, length(labs)))
      }
    } else {
      stop("if labs is unspecified then the first column of scores must be character data")
    }
  }

  
  
  # If NA is supplied then put blank labels  
  if(all(is.na(labs))) {
    labs <- rep("", length(scores[[1]]))
  }
    
  # Check data is in right shape
  x <- vapply(scores, function(x) length(x)==length(labs), FALSE)
  if(!all(x)) {
    stop("Each score vector must be the same length as the labs vector")
  }
  
  colMatrix <- colourMatrix(colMatrix)
  
  # Check for maxScale
  opScale <- list(scale = list())
  opScale$scale <- setRadarScale(maxScale, scaleStepWidth, scaleStartValue)
  opScale$scale$pointLabels <- list(fontSize = labelSize) 
                                   #fontColor = "#111",
                                   #fontFamily = "Times")
  
  opToolTip <- list(tooltips = list(enabled = showToolTipLabel,
                                    mode = "label"))
  
  # Any extra options passed straight through. Names must match existing options
  # http://www.chartjs.org/docs/#getting-started-global-chart-configuration
  opPassThrough <- list(...)
  
  opTitle <- list(title = list(display = !is.null(main), text = main))

  opLegend <- list(legend = list(display = showLegend))
  
  # Combine scale options, pass through and explicit options
  opList <- c(list(responsive = responsive),
              opTitle, opScale, opToolTip, opLegend, opPassThrough)
              
    
  # forward options using x
  datasets <- lapply(names(scores), function(x) list(label=x))
  
  for (i in seq_along(datasets)) {
    
    iCol <- (i-1) %% ncol(colMatrix) + 1 # cyclic repeat colours

    fillCol <- paste0("rgba(", paste0(colMatrix[ , iCol], collapse=","),
                      ",", polyAlpha, ")")
    lineCol <- paste0("rgba(", paste0(colMatrix[ , iCol], collapse=","), 
                      ",", lineAlpha, ")")
    
    datasets[[i]]$data <- scores[[i]]             # Data Points
    
    datasets[[i]]$backgroundColor  <- fillCol           # Polygon Fill
    datasets[[i]]$borderColor  <- lineCol         # Line Colour
    datasets[[i]]$pointBackgroundColor  <- lineCol          # Point colour
    
    datasets[[i]]$pointBorderColor  <- "#fff"     # Point outline
    datasets[[i]]$pointHoverBackgroundColor  <- "#fff"   # Point Highlight fill
    datasets[[i]]$pointHoverBorderColor <- lineCol # Point Highlight line
    if(!addDots) datasets[[i]]$pointRadius <- 0
  }
  
  x <- list(data = list(labels=labs, datasets=datasets), options = opList)
  
  #print(jsonlite::toJSON(x, pretty = TRUE, auto_unbox = TRUE))
  
  # create widget
  htmlwidgets::createWidget(
    name = "chartJSRadar",
    x,
    width = width,
    height = height,
    package = "radarchart"
  )
}

#' Autoscale the radar plot
#'
#' @param maxScale Numeric length 1. Desired max limit
#' @param scaleStepWidth Numeric length 1. Spacing between rings
#' @param scaleStartValue Numeric length 1. Value of the centre
#'
#' @return A list containing the scale options for chartjs
#'
#' @examples
#' \dontrun{
#' setRadarScale(15, 3)
#' setRadarScale(15, 5, 2)
#' }
setRadarScale <- function(maxScale = NULL, scaleStepWidth = NULL, 
                          scaleStartValue = 0) {
  
  if (!is.null(maxScale)) {
    
    opScale <- list(ticks = list(max = maxScale))
    opScale$ticks$min <- scaleStartValue
    
    # Did they fix the tick points?
    if (!is.null(scaleStepWidth)) {
      opScale$ticks$stepSize <- scaleStepWidth
      opScale$ticks$maxTicksLimit <- 1000
    }
    # else {
    #   if (maxScale-scaleStartValue <= 12) {
    #     opScale$scaleStepWidth <- 1
    #   } else {
    #     opScale$scaleStepWidth <- floor( (maxScale-scaleStartValue) / 10)
    #   }
    # }
    # opScale$scaleSteps <- ceiling( (maxScale - scaleStartValue) / 
    #                                 opScale$scaleStepWidth)
    # opScale$scaleStartValue <- scaleStartValue
  } else {
    if(!is.null(scaleStartValue)) {
      opScale <- list(ticks = list(min = scaleStartValue))
    } else {
      opScale <- list()
    }
  }
  opScale
}

#' Tell htmltools where to output the chart
#' 
#' @param id The id of the target object
#' @param style css stylings
#' @param class class of the target
#' @param width width of target
#' @param height height of target
#' @param ... extra arguments currently unused
#' 
#' @export
chartJSRadar_html <- function(id, style, class, width, height, ...){
  htmltools::tags$canvas(id = id, class = class, width=width, height=height)
}

#' Widget output function for use in Shiny
#'
#' @param outputId output variable to read from
#' @param width Must be valid CSS unit
#' @param height Must be valid CSS unit
#'
#' @export
chartJSRadarOutput <- function(outputId, width = "450", height = "300"){
  shinyWidgetOutput(outputId, "chartJSRadar", width, height, package = "radarchart")
}

#' Widget render function for use in Shiny
#'
#' @param expr expression passed to \link[htmlwidgets]{shinyRenderWidget}
#' @param env environment in which to evaluate expression
#' @param quoted Logical. Is expression quoted?
#'
#' @export
renderChartJSRadar <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { 
    expr <- substitute(expr) 
  } # force quoted
  shinyRenderWidget(expr, chartJSRadarOutput, env, quoted = TRUE)
}
